### PR TITLE
Fix negative indexes and check out-of-bound indexes for vindex

### PR DESCRIPTION
**Pull request by shoyer**
- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -3043,10 +3043,19 @@ def _vindex(x, *indexes):
...
+        if not isinstance(ind, slice):
+            ind = np.array(ind)
+            if ind.dtype.kind == 'b':
+                raise IndexError('vindex does not support indexing with '
+                                 'boolean arrays')
+            if ((ind >= size) | (ind < -size)).any():
+                raise IndexError('vindex key has entries out of bounds for '
+                                 'indexing along axis %s of size %s: %r'
+                                 % (i, size, ind))
+            ind %= size
```
Comment (mrocklin) [inline dask/array/core.py]: Is `ind` definitely a copy here?  Is there some chance that we are mutating a user input?

**Comment by mrocklin** [comment]
This generally looks good to me.

File: dask/array/core.py (hunk 1 above)
Comment (shoyer) [reply dask/array/core.py]: Yes, `np.array()` makes a copy by default. But I'll call `copy=True` explicitly to make that obvious.

**Comment by mrocklin** [comment]
Merging later today if there are no further comments

File: dask/array/core.py [hunk 2]
```diff
@@ -3043,10 +3043,19 @@ def _vindex(x, *indexes):
...
+            ind = np.array(ind, copy=True)
+            if ind.dtype.kind == 'b':
+                raise IndexError('vindex does not support indexing with '
+                                 'boolean arrays')
+            if ((ind >= size) | (ind < -size)).any():
+                raise IndexError('vindex key has entries out of bounds for '
+                                 'indexing along axis %s of size %s: %r'
+                                 % (i, size, ind))
+            ind %= size
+            array_indexes[i] = ind
```
Comment (pitrou) [inline dask/array/core.py]: I'm curious: is `np.broadcast_arrays(*array_indexes.values())` relying on the *ordering* of dict values?

Comment (shoyer) [reply dask/array/core.py]: It only relies on dict order being consistent between `.values()` and `.keys()`, not any particular order for dict iteration.

**Comment by mrocklin** [comment]
OK, merging.  Thanks @shoyer

**Pull request merged**